xds/internal/xdsclient/pool.go

```diff
@@ -0,0 +1,143 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/xds/bootstrap"
+)
+
+var (
+	// DefaultPool is the default pool for xds clients. It is created at the
+	// init time.
+	DefaultPool *Pool
+)
+
+// Pool represents pool of xds clients.
+type Pool struct {
+	mu      sync.Mutex
+	clients map[string]*clientRefCounted
+	config  *bootstrap.Config
+}
+
+func init() {
+	DefaultPool = &Pool{clients: make(map[string]*clientRefCounted)}
+}
+
+// NewPool creates a new xds client pool with the given bootstrap contents.
+func NewPool(bootstrapContents []byte) (*Pool, error) {
```

Comment:
I think it would be nicer to have this accept a `*bootstrap.Config` (which could optionally be `nil`) rather than having to parse the configuration here. With that change, this API will return a single value, `*Pool`, instead of two.

Reply:
When we change this to accept a `bootstrap.Config`, we should also export `bootstrap.newConfigFromContents`. This would allow non-test usages for `NewPool` where the caller would use the exported `bootstrap.newConfigFromContents` to parse the JSON config and pass it to this function.

Reply:
> I think it would be nicer to have this accept a *bootstrap.Config (which could optionally be nil) rather than having to parse the configuration here. With that change, this API will return a single value, *Pool, instead of two.

Done

Reply:
> we should also export bootstrap.newConfigFromContents. This would allow non-test usages for NewPool

We already have `bootstrap.NewConfigForTesting` which internally call `bootstrap.newConfigFromContents`